### PR TITLE
fix bug for issue #1711

### PR DESCRIPTION
**Pull request by alpha-baby**
### Issues associated with this PR

#1711

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/upstream/cluster/loadbalancer.go [hunk 1]
```diff
@@ -298,8 +298,12 @@ func (lb *EdfLoadBalancer) ChooseHost(context types.LoadBalancerContext) types.H
 		return nil
 	}
 	if total == 1 {
+		targetHost := targetHosts[0]
 		// Return directly if there is only one host
-		return targetHosts[0]
+		if targetHost.Health() {
+			return targetHost
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/loadbalancer.go:304]: should return nil when all hosts are unhealthy?

Comment (alpha-baby) [reply pkg/upstream/cluster/loadbalancer.go:304]: 我也想过这个问题，因为choosehost这个方法就是定义的返回值只有一个，如果在调用choosehost的时候返回了nil就说明出了问题

按照go的哲学我觉得可以再定义一个error的返回值，这样就可以告诉调用者是没有host还是说，host都不健康了，还是说根据规则没有匹配到host。

我觉得可以定义几个标准的error，例如：

1. ErrNoHost
2. ErrUnavailableHost
3. ErrNotMatchedHost

Comment (taoyuanyuan) [reply pkg/upstream/cluster/loadbalancer.go:304]: 我觉得可以， @nejisama  看看呢

Comment (nejisama) [reply pkg/upstream/cluster/loadbalancer.go:304]: 返回不同的error，然后准备针对不同的error 做什么特殊的策略吗？ 仅仅是为了日志的话，直接在LB里加日志就可以了吧

Comment (alpha-baby) [reply pkg/upstream/cluster/loadbalancer.go:304]: 这个标准错误可以返回给接入的应用

Comment (taoyuanyuan) [reply pkg/upstream/cluster/loadbalancer.go:304]: 这个问题我们可以新开issue看看， 这个issue先看看另外一个问题，就是一个不健康和所有不健康的逻辑不一致，所有不健康是随机选一个，一个不健康直接返回nil，逻辑上冲突了

Comment (taoyuanyuan) [reply pkg/upstream/cluster/loadbalancer.go:304]: https://github.com/mosn/mosn/blob/b31d102f35544b31c7abe3baadbb6808a8691fef/pkg/upstream/cluster/loadbalancer.go#L319

Comment (taoyuanyuan) [reply pkg/upstream/cluster/loadbalancer.go:304]: 嗯 那这儿所有不健康也返回nil

File: pkg/upstream/cluster/loadbalancer.go [hunk 2]
```diff
@@ -320,8 +320,9 @@ func (lb *EdfLoadBalancer) ChooseHost(context types.LoadBalancerContext) types.H
 	}
 	lb.mutex.Lock()
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/loadbalancer.go:321]: 锁删了

**Review by taoyuanyuan — APPROVED** [review]
LGTM

**Review by nejisama — APPROVED** [review]
LGTM
